src/cunumeric/stat/histogram_cpu.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/system/omp/execution_policy.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <tuple>
+#include <vector>
+
+#include "cunumeric/stat/histogram_gen.h"
+
+namespace cunumeric {
+namespace detail {
+template <typename element_t>
+decltype(auto) get_raw_ptr(Buffer<element_t>& v)
+{
+  return v.ptr(0);
+}
+
+// host specialization
+//
+template <typename elem_t, typename exe_policy_t>
+struct allocator_t<
+  elem_t,
+  exe_policy_t,
+  std::enable_if_t<std::is_same_v<exe_policy_t, thrust::detail::host_t> ||
+                   std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
+  allocator_t(void) {}
+
+  allocator_t(elem_t, exe_policy_t) {}  // tag-dispatch for CTAD
+
+  elem_t* operator()(exe_policy_t exe_pol, size_t size, elem_t init = 0)
+  {
+    d_buffer_     = create_buffer<elem_t>(size);
+    elem_t* d_ptr = get_raw_ptr(d_buffer_);
+
+    std::fill_n(d_ptr, size, init);
+
+    return d_ptr;
+  }
+
+ private:
+  Buffer<elem_t> d_buffer_;
+};
+
+template <typename exe_policy_t>
+void synchronize_exec(exe_policy_t, cudaStream_t stream)
+{
+  // nothing;
+}
+
+// host specialization:
+//
+template <typename exe_policy_t, typename weight_t, typename offset_t, typename allocator_t>
+struct segmented_sum_t<
+  exe_policy_t,
+  weight_t,
+  offset_t,
+  allocator_t,
+  std::enable_if_t<std::is_same_v<exe_policy_t, thrust::detail::host_t> ||
+                   std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
+  segmented_sum_t(exe_policy_t exe_pol,
+                  weight_t const* p_weights,
+                  size_t n_samples,
+                  weight_t* p_hist,
+                  size_t n_intervals,
+                  offset_t* p_offsets,
+                  cudaStream_t stream,
+                  allocator_t& allocator)
+    : ptr_weights_(p_weights),
+      n_samples_(n_samples),
+      ptr_hist_(p_hist),
+      n_intervals_(n_intervals),
+      ptr_offsets_(p_offsets)
+  {
+  }
+  void operator()(void)
+  {
+    for (auto interval_index = 0; interval_index < n_intervals_; ++interval_index) {
+      auto next_index = interval_index + 1;
+
+      ptr_hist_[interval_index] = std::accumulate(
```

Comment:
I can try. Notice, though, this is already called from a potential parallel loop (in the OMP version). So, this would trigger the OMP equivalent of dynamic parallelism, whereby you call an algorithm with a parallel execution policy from within an parallel thread (in this case OMP). Even if that works, it might bring a performance penalty (it would on device).

But, I can try calling `thrust::reduce(thrust::seq,...)`; i.e., w/ `sequential` execution policy. This should work.

Reply:
Ok, done.